File: scripts/Morse/Morse_2D/exact_eigenvalues/2Dmorse_theory.f90
!=============================================================================80
!                           2D Morse Eigenvalues
!=============================================================================80
!Compute Eigenvalues for the 2D Morse Oscillator (Analytic Expression)
!The eigenvalues will become negative due to the decreasing distance between 
!energy states at higher excitations (After 24 excitations in 1D). 
!==============================================================================!
!The energies output (theory.dat) contain all the combinatorics 
!These values are NOT sorted, see Python script "sort_2Dmorse.py" for sorting
!==============================================================================!
!       Modified:
!   5 May 2019
!       Author:
!   Shane Flynn
!==============================================================================!
program main
!==============================================================================!
!NB             ==>Number of Excitations (Number of Basis Functions)
!ax,ay          ==>Parameters Taken from JCP Garaschuk/Light 2001
!Dmorse         ==>Morse Parameter
!==============================================================================!
implicit none
integer::i,j,NB
double precision::D_morse,ax,ay
double precision,allocatable::eigx(:),eigy(:)
!==============================================================================!
!E(n):=a*sqrt(2*D_morse)*(n+1/2)-((a*sqrt(2*D_morse)*(n+1/2))^2/(4*D_morse))
!==============================================================================!
D_morse=12
ax=0.2041241  
ay=0.18371169 
NB=24                            !will become negative with more basis functions
!==============================================================================!
open(unit=17,file='1Dmorse_eig.dat')
allocate(eigx(NB),eigy(NB))
!==============================================================================!
!Compute 1D excitations, Excitaitons start at E_0 (ground state)
!==============================================================================!
do i=0,NB-1      
   eigx(i+1)=ax*sqrt(2.*D_morse)*(i+0.5)&
       - ((ax*sqrt(2.*D_morse)*(i+0.5))**2/(4.*D_morse))
   eigy(i+1)=ay*sqrt(2.*D_morse)*(i+0.5)&
       - ((ay*sqrt(2.*D_morse)*(i+0.5))**2/(4.*D_morse))
   write(17,*) eigx(i+1), eigy(i+1)
enddo
close(17)
!==============================================================================!
!Compute Combinatorics. Potential is Seperable i.e. E_total=E_x+E_y
!==============================================================================!
open(unit=90,file='2Dmorse_eig.dat')
do i=1,NB
    do j=1,NB
        write(90,*) eigx(i)+eigy(j)
    enddo
enddo
close(90)
write(*,*) 'Hello Universe!'
end program main

File: scripts/Morse/Morse_2D/exact_eigenvalues/2Dmorse_theory.py
#=============================================================================80
#                           2D Morse Eigenvalues
#=============================================================================80
#Compute Eigenvalues for the 2D Morse Oscillator (Analytic Expression)
#The eigenvalues will become negative due to the decreasing distance between 
#energy states at higher excitations (After 24 excitations in 1D). 
#==============================================================================#
#current code gives different results compared to fortran after ~6 decimals, 
#need to figure out why there is a descrepancy, Use this code at your own risk!!
#==============================================================================#
#       Modified:
#   5 May 2019
#       Author:
#   Shane Flynn 
#==============================================================================#
import pandas as pd
import numpy as np
#==============================================================================#
#                               2D Morse Parameters
#==============================================================================#
D_morse=12
ax=0.2041241  
ay=0.18371169 
NB=23
#==============================================================================#
#           Compute 1D eigenvalues for x,y dimensions seperately
#==============================================================================#
eigx=[]
eigy=[]
eigs=[]
for i in range(0,NB):
    eigx.append(np.float(ax*np.sqrt(2.*D_morse)*(i+0.5)- \
            ((ax*np.sqrt(2.*D_morse)*(i+0.5))**2/(4.*D_morse))))
    eigy.append(np.float(ay*np.sqrt(2.*D_morse)*(i+0.5)- \
            ((ay*np.sqrt(2.*D_morse)*(i+0.5))**2/(4.*D_morse))))
print eigx
print eigy
#==============================================================================#
#                       Compute Excitation Combinatorics
#==============================================================================#
for i in range(0,NB):
    for j in range(0,NB):
        eigs.append(np.float(eigx[i]+eigy[j]))
#==============================================================================#
#                              Sort by Magnitude
#==============================================================================#
df_theory=pd.DataFrame(eigs)
df_sort=df_theory.sort_values([0])
df_sort=df_sort.reset_index(drop=True)
#==============================================================================!
#                           Write Data to CSV File
#==============================================================================!
df_sort.to_csv('sort_theory.dat',sep='\t',index=False)
